From 8ae02631a9806da11b34cd6b274af02d28aee5da Mon Sep 17 00:00:00 2001
From: Eli Schwartz <eschwartz@archlinux.org>
Date: Thu, 26 Mar 2020 13:49:45 -0400
Subject: [PATCH 20/20] Replace all remaining usage of non-portable 'which'
 with POSIX command -v

The "which" utility is not guaranteed to be installed, and if it
is, its behavior is not portable.

Conversely, the "command -v" shell builtin is required to exist in all
POSIX 2008 compliant shells, and is thus guaranteed to work everywhere.

Examples of open-source shells likely to be installed as /bin/sh on
Linux, which implement the 12-year-old standard: ash, bash, busybox,
dash, ksh, mksh and zsh.

A side benefit of using the POSIX portable option is that it does not
require an external disk executable to be forked. This therefore
represents a mild speedup.
---
 scripts/xdg-email.in              |  8 ++++----
 scripts/xdg-file-dialog.in        | 16 ++++++++--------
 scripts/xdg-mime.in               |  6 +++---
 scripts/xdg-open.in               |  2 +-
 scripts/xdg-screensaver.in        |  4 ++--
 scripts/xdg-su.in                 | 14 +++++++-------
 scripts/xdg-terminal.in           | 14 +++++++-------
 scripts/xdg-utils-common.in       |  8 ++++----
 tests/include/desktop_environment |  9 +++------
 tests/include/system_info         |  3 +--
 tests/testrun                     |  4 ++--
 11 files changed, 42 insertions(+), 46 deletions(-)

diff --git a/scripts/xdg-email.in b/scripts/xdg-email.in
index 85cf90a..dfb4b83 100644
--- a/scripts/xdg-email.in
+++ b/scripts/xdg-email.in
@@ -90,7 +90,7 @@ open_kde()
         local kreadconfig=kreadconfig
     fi
 
-    if which $kreadconfig >/dev/null 2>&1; then
+    if command -v $kreadconfig >/dev/null; then
         local profile=$($kreadconfig --file emaildefaults \
                                      --group Defaults --key Profile)
         if [ -n "$profile" ]; then
@@ -98,7 +98,7 @@ open_kde()
                                         --group "PROFILE_$profile" \
                                         --key EmailClient \
                                   | cut -d ' ' -f 1)
-            if [ -z "${client%%*.desktop}" ] && ! which "$client" >/dev/null 2>&1; then
+            if [ -z "${client%%*.desktop}" ] && ! command -v "$client" >/dev/null 2>&1; then
                 client=`desktop_file_to_binary "$client"`
             fi
 
@@ -115,7 +115,7 @@ open_kde()
         *) command=kde-open$KDE_SESSION_VERSION ;;
     esac
 
-    if which $command >/dev/null 2>&1; then
+    if command -v $command >/dev/null; then
         DEBUG 1 "Running $command \"$1\""
         if [ "$KDE_SESSION_VERSION" = 3 ]; then
             # KDE3 uses locale's encoding when decoding the URI,
@@ -435,7 +435,7 @@ mailto=`echo "${mailto}"| sed 's/[?&]$//'`
 # Shouldn't happen
 [ x"${mailto}" != x"" ] || exit_failure_syntax
 
-if which @NAME@-hook.sh > /dev/null 2> /dev/null; then
+if command -v @NAME@-hook.sh > /dev/null; then
     @NAME@-hook.sh "${mailto}"
     if [ $? -eq 0 ]; then
         exit_success
diff --git a/scripts/xdg-file-dialog.in b/scripts/xdg-file-dialog.in
index 2b17fd5..0db2672 100644
--- a/scripts/xdg-file-dialog.in
+++ b/scripts/xdg-file-dialog.in
@@ -31,7 +31,7 @@ _USAGE
 
 open_kde()
 {
-    DIALOG=`which kdialog`
+    DIALOG=`command -v kdialog`
     if [ $? -eq 0 ]; then
         if [ x"$TITLE" != x"" ]; then
             $DIALOG --title "$TITLE" --getopenfilename "$1" ""
@@ -51,7 +51,7 @@ open_kde()
 
 open_zenity()
 {
-    DIALOG=`which zenity`
+    DIALOG=`command -v zenity`
     if [ $? -eq 0 ]; then
         if [ x"$1" != x"" ]; then
             cd `dirname "$1"` 2>/dev/null
@@ -87,7 +87,7 @@ open_zenity()
 
 open_multi_kde()
 {
-    DIALOG=`which kdialog`
+    DIALOG=`command -v kdialog`
     if [ $? -eq 0 ]; then
         if [ x"$TITLE" != x"" ]; then
             $DIALOG --title "$TITLE" --multiple --separate-output \
@@ -108,7 +108,7 @@ open_multi_kde()
 
 open_multi_zenity()
 {
-    DIALOG=`which zenity`
+    DIALOG=`command -v zenity`
     if [ $? -eq 0 ]; then
         if [ x"$1" != x"" ]; then
             cd `dirname "$1"` 2>/dev/null
@@ -145,7 +145,7 @@ open_multi_zenity()
 
 save_kde()
 {
-    DIALOG=`which kdialog`
+    DIALOG=`command -v kdialog`
     if [ $? -eq 0 ]; then
         if [ x"$TITLE" != x"" ]; then
             $DIALOG --title "$TITLE" --getsavefilename "$1" ""
@@ -165,7 +165,7 @@ save_kde()
 
 save_zenity()
 {
-    DIALOG=`which zenity`
+    DIALOG=`command -v zenity`
     if [ $? -eq 0 ]; then
         if [ x"$1" != x"" ]; then
             cd `dirname "$1"` 2>/dev/null
@@ -201,7 +201,7 @@ save_zenity()
 
 directory_kde()
 {
-    DIALOG=`which kdialog`
+    DIALOG=`command -v kdialog`
     if [ $? -eq 0 ]; then
         if [ x"$TITLE" != x"" ]; then
             $DIALOG --title "$TITLE" --getexistingdirectory "$1" ""
@@ -221,7 +221,7 @@ directory_kde()
 
 directory_zenity()
 {
-    DIALOG=`which zenity`
+    DIALOG=`command -v zenity`
     if [ $? -eq 0 ]; then
         if [ x"$1" != x"" ]; then
             cd "$1" 2>/dev/null
diff --git a/scripts/xdg-mime.in b/scripts/xdg-mime.in
index 612d2ce..d194b0e 100644
--- a/scripts/xdg-mime.in
+++ b/scripts/xdg-mime.in
@@ -441,14 +441,14 @@ defapp_kde()
     if [ -n "${KDE_SESSION_VERSION}" ]; then
       case "${KDE_SESSION_VERSION}" in
         4)
-          KTRADER=`which ktraderclient 2> /dev/null`
+          KTRADER=`command -v ktraderclient`
         ;;
         5)
-          KTRADER=`which ktraderclient${KDE_SESSION_VERSION} 2> /dev/null`
+          KTRADER=`command -v ktraderclient${KDE_SESSION_VERSION}`
         ;;
       esac
     else
-        KTRADER=`which ktradertest 2> /dev/null`
+        KTRADER=`command -v ktradertest`
     fi
     if [ -n "$KTRADER" ] ; then
         DEBUG 1 "Running KDE trader query \"$MIME\" mimetype and \"Application\" servicetype"
diff --git a/scripts/xdg-open.in b/scripts/xdg-open.in
index 9fc3ff8..50e31e6 100644
--- a/scripts/xdg-open.in
+++ b/scripts/xdg-open.in
@@ -411,7 +411,7 @@ open_generic()
             open_generic_xdg_mime "$file" "$filetype"
         fi
 
-        if which run-mailcap 2>/dev/null 1>&2; then
+        if command -v run-mailcap >/dev/null; then
             run-mailcap --action=view "$file"
             if [ $? -eq 0 ]; then
                 exit_success
diff --git a/scripts/xdg-screensaver.in b/scripts/xdg-screensaver.in
index 95f707e..c768f91 100644
--- a/scripts/xdg-screensaver.in
+++ b/scripts/xdg-screensaver.in
@@ -38,7 +38,7 @@ else
    MV="mv"
    screensaver_file="$HOME/.xdg-screensaver-"`echo $(hostname)-$DISPLAY | sed 's/:/-/g'`
 fi
-lockfile_command=`which lockfile 2> /dev/null`
+lockfile_command=`command -v lockfile`
 
 lockfile()
 {
@@ -166,7 +166,7 @@ do_resume()
   cleanup_suspend
 }
 
-XPROP=`which xprop 2> /dev/null`
+XPROP=`command -v xprop`
 
 check_window_id()
 {
diff --git a/scripts/xdg-su.in b/scripts/xdg-su.in
index 41b3109..c6b469d 100644
--- a/scripts/xdg-su.in
+++ b/scripts/xdg-su.in
@@ -36,7 +36,7 @@ su_kde()
     if [ x"$KDE_SESSION_VERSION" = x"4" ]; then
         KDESU=`kde4-config --locate kdesu --path exe 2>/dev/null`
     else
-        KDESU=`which kdesu 2>/dev/null`
+        KDESU=`command -v kdesu`
     fi
     if [ $? -eq 0 ] ; then
         if [ -z "$user" ] ; then
@@ -57,9 +57,9 @@ su_kde()
 
 su_gnome()
 {
-    GSU=`which gnomesu 2>/dev/null`
+    GSU=`command -v gnomesu`
     if [ $? -ne 0 ] ; then
-        GSU=`which xsu 2>/dev/null`
+        GSU=`command -v xsu`
     fi
     if [ $? -eq 0 ] ; then
         if [ -z "$user" ] ; then
@@ -80,7 +80,7 @@ su_gnome()
 
 su_lxqt()
 {
-    LXQTSU=`which lxqt-sudo 2>/dev/null`
+    LXQTSU=`command -v lxqt-sudo`
     if [ $? -eq 0 ] ; then
         if [ -z "$user" ] ; then
              # -s option runs as su rather then sudo
@@ -104,7 +104,7 @@ su_enlightenment()
 {
 # Enlightenment doesn't have any reasonably working su/sudo graphical interface
 # but terminology works as a drop in replacement for xterm and has a matching theme
-    if which terminology >/dev/null ; then
+    if command -v terminology >/dev/null ; then
         if [ -z "$user" ] ; then
             terminology -g 60x5 -T "xdg-su: $cmd" -e "su -c '$cmd'"
         else
@@ -138,7 +138,7 @@ su_generic()
 
 su_xfce()
 {
-    if which gnomesu >/dev/null 2>&1 ; then
+    if command -v gnomesu >/dev/null ; then
         su_gnome
     else
         su_generic
@@ -187,7 +187,7 @@ fi
 detectDE
 
 if [ x"$DE" = x"" ]; then
-    XSU=`which xterm 2>/dev/null`
+    XSU=`command -v xterm`
     if [ $? -eq 0 ] ; then
         DE=generic
     fi
diff --git a/scripts/xdg-terminal.in b/scripts/xdg-terminal.in
index caefa0d..71d56a5 100644
--- a/scripts/xdg-terminal.in
+++ b/scripts/xdg-terminal.in
@@ -32,7 +32,7 @@ terminal_kde()
 {
     terminal=`kreadconfig --file kdeglobals --group General --key TerminalApplication --default konsole`
 
-    terminal_exec=`which $terminal 2>/dev/null`
+    terminal_exec=`command -v $terminal`
 
     if [ -x "$terminal_exec" ]; then
         if [ x"$1" = x"" ]; then
@@ -59,7 +59,7 @@ terminal_gnome()
     term_exec=`gconftool-2 --get ${term_exec_key}`
     term_exec_arg=`gconftool-2 --get ${term_exec_arg_key}`
 
-    terminal_exec=`which $term_exec 2>/dev/null`
+    terminal_exec=`command -v $term_exec`
 
     if [ -x "$terminal_exec" ]; then
         if [ x"$1" = x"" ]; then
@@ -95,7 +95,7 @@ terminal_gsettings()
     term_exec=`gsettings get ${term_schema} exec | sed -r "s/^'(.*)'$/\1/"`
     term_exec_arg=`gsettings get ${term_schema} exec-arg | sed -r "s/^'(.*)'$/\1/"`
 
-    terminal_exec=`which $term_exec 2>/dev/null`
+    terminal_exec=`command -v $term_exec`
 
     if [ -x "$terminal_exec" ]; then
         if [ x"$1" = x"" ]; then
@@ -146,7 +146,7 @@ terminal_generic()
         TERM=xterm
     fi
 
-    terminal_exec=`which $TERM 2>/dev/null`
+    terminal_exec=`command -v $TERM`
 
     if [ -x "$terminal_exec" ]; then
         if [ x"$1" = x"" ]; then
@@ -177,7 +177,7 @@ terminal_generic()
 
 terminal_lxde()
 {
-    if which lxterminal >/dev/null 2>&1; then
+    if command -v lxterminal >/dev/null; then
         if [ x"$1" = x"" ]; then
             lxterminal
         else
@@ -190,7 +190,7 @@ terminal_lxde()
 
 terminal_lxqt()
 {
-    if which qterminal >/dev/null 2>&1; then
+    if command -v qterminal >/dev/null; then
         if [ x"$1" = x"" ]; then
             qterminal
         else
@@ -203,7 +203,7 @@ terminal_lxqt()
 
 terminal_enlightenment()
 {
-    if which terminology >/dev/null 2>&1; then
+    if command -v terminology >/dev/null; then
         if [ x"$1" = x"" ]; then
             terminology
         else
diff --git a/scripts/xdg-utils-common.in b/scripts/xdg-utils-common.in
index ed6c337..0147527 100644
--- a/scripts/xdg-utils-common.in
+++ b/scripts/xdg-utils-common.in
@@ -23,7 +23,7 @@ first_word()
 binary_to_desktop_file()
 {
     search="${XDG_DATA_HOME:-$HOME/.local/share}:${XDG_DATA_DIRS:-/usr/local/share:/usr/share}"
-    binary="`which "$1"`"
+    binary="`command -v "$1"`"
     binary="`readlink -f "$binary"`"
     base="`basename "$binary"`"
     IFS=:
@@ -38,7 +38,7 @@ binary_to_desktop_file()
             # Make sure it's a visible desktop file (e.g. not "preferred-web-browser.desktop").
             grep -Eq "^(NoDisplay|Hidden)=true" "$file" && continue
             command="`grep -E "^Exec(\[[^]=]*])?=" "$file" | cut -d= -f 2- | first_word`"
-            command="`which "$command"`"
+            command="`command -v "$command"`"
             if [ x"`readlink -f "$command"`" = x"$binary" ]; then
                 # Fix any double slashes that got added path composition
                 echo "$file" | sed -e 's,//*,/,g'
@@ -80,7 +80,7 @@ desktop_file_to_binary()
         if [ -r "$file_path" ]; then
             # Remove any arguments (%F, %f, %U, %u, etc.).
             command="`grep -E "^Exec(\[[^]=]*])?=" "$file_path" | cut -d= -f 2- | first_word`"
-            command="`which "$command"`"
+            command="`command -v "$command"`"
             readlink -f "$command"
             return
         fi
@@ -355,7 +355,7 @@ detectDE()
     if [ x"$DE" = x"gnome" ]; then
       # gnome-default-applications-properties is only available in GNOME 2.x
       # but not in GNOME 3.x
-      which gnome-default-applications-properties > /dev/null 2>&1  || DE="gnome3"
+      command -v gnome-default-applications-properties > /dev/null || DE="gnome3"
     fi
 
     if [ -f "$XDG_RUNTIME_DIR/flatpak-info" ]; then
diff --git a/tests/include/desktop_environment b/tests/include/desktop_environment
index 308906b..22bfd2d 100755
--- a/tests/include/desktop_environment
+++ b/tests/include/desktop_environment
@@ -2,8 +2,7 @@
 
 ## This script modified from one supplied by Bryce Harrington at OSDL
 
-which kde-config >/dev/null 2>&1
-if [ "$?" -eq 0 ]; then
+if command -v kde-config >/dev/null; then
 	# From Sutoka on FreeNode/#kde
 	kde_version=`kde-config --version | grep KDE | cut -d' ' -f2 2> /dev/null`
 	kde_running=$KDE_FULL_SESSION
@@ -16,8 +15,7 @@ else
 fi
 
 # From kees on #osdl
-which gnome-session >/dev/null 2>&1
-if [ "$?" -eq 0 ] ; then
+if command -v gnome-session >/dev/null ; then
     gnome_version=`gnome-session --version | cut -d ' ' -f3 2>/dev/null`
     echo "gnome: $gnome_version"
 
@@ -28,8 +26,7 @@ else
     echo "gnome: not present (didn't find gnome-session)"
 fi
 
-which xfce4-session >/dev/null 2>&1
-if [ "$?" -eq 0 ] ; then 
+if command -v xfce4-session >/dev/null ; then
         xfce_version=`xfce4-session --version | grep '(Xfce ' | cut -d '(' -f 2 | cut -d ' ' -f2 | cut -d ')' -f1`
 	# From massonnet on FreeNode/#xfce
 	#if [ `pidof xfce4-session` ]; then
diff --git a/tests/include/system_info b/tests/include/system_info
index b689a30..f67d69b 100755
--- a/tests/include/system_info
+++ b/tests/include/system_info
@@ -1,8 +1,7 @@
 #!/bin/bash
 
 parentdir=${0%/*}
-which perl >/dev/null 2>&1
-if [ "$?" ] ; then
+if command -v perl >/dev/null ; then
 	distro=`perl $parentdir/linux_distro 2>&1`
 else
 	distro="not available (couldn't find perl)"
diff --git a/tests/testrun b/tests/testrun
index c169f73..11ec55d 100755
--- a/tests/testrun
+++ b/tests/testrun
@@ -75,7 +75,7 @@ if [ -z "$XDG_TEST_DIR" ] ; then
 	export XDG_TEST_DIR="$PWD"
 	echo "WARNING: guessed XDG_TEST_DIR to be $XDG_TEST_DIR"
 fi
-if [ -z `which xdg-mime 2>/dev/null` ] && [ -d "$XDG_TEST_DIR/../scripts" ] ; then
+if ! command -v xdg-mime >/dev/null && [ -d "$XDG_TEST_DIR/../scripts" ] ; then
 	export PATH="$PATH:$XDG_TEST_DIR/../scripts"
 	echo "WARNING: modified PATH to add '$XDG_TEST_DIR/../scripts'"
 fi
@@ -121,7 +121,7 @@ if [ -z "$XDG_TEST_SELF_LAUNCH"  ] ; then  # not self launched
 			echo "Running su for system tests. (Use testrun -S to use sudo instead)"
 			echo "Please enter the root password when requested."
 		else
-			SUCMD=`which sudo 2>/dev/null`
+			SUCMD=`command -v sudo`
 			echo "Running ${SUCMD-su} for system tests."
 			echo "Please enter an appropriate password if requested."
 		fi
-- 
2.38.1

